src/stream.iced

```diff
@@ -0,0 +1,114 @@
+stream = require('keybase-chunk-stream')
+nacl = require('keybase-nacl')
+armor = require('node-armor-x')
+msgpack = require('msgpack-lite')
+payload = require('./payload')
+header = require('./header')
+nonce = require('./nonce')
+format = require('./format')
+
+saltpack_block_len = (1024**2)
+noop = () ->
+
+class NaClEncryptStream extends stream.ChunkStream
+  _encrypt : (chunk) =>
+    payload_list = payload.generate_encryption_payload_packet(@_encryptor, chunk, @_block_num, @_header_hash, @_mac_keys)
+    ++@_block_num
+    return payload_list
+
+  _transform : (chunk, encoding, cb) ->
+    if not @_header_written
+      {header_intermediate, header_hash, mac_keys, payload_key} = header.generate_encryption_header_packet(@_encryptor, @_recipients, {anonymized_recipients : @_anonymized_recipients})
+      @_header_hash = header_hash
+      @_mac_keys = mac_keys
+      @_encryptor.secretKey = payload_key
+      @push(header_intermediate)
+      @_header_written = true
+    super(chunk, encoding, cb)
+
+  _flush : (cb) ->
+    super(noop)
+    @push(@_encrypt(Buffer.from('')))
+    cb()
+
+  constructor : (@_encryptor, @_recipients, @_anonymized_recipients) ->
+    @_header_written = false
+    @_block_num = 0
+    @_mac_keys = null
+    @_header_hash = null
+    super(@_encrypt, {block_size : saltpack_block_len, exact_chunking : true, writableObjectMode : false, readableObjectMode : true})
+
+class NaClDecryptStream extends stream.ChunkStream
+  _decrypt : (chunk) =>
+    payload_text = payload.parse_encryption_payload_packet(@_decryptor, chunk, @_block_num, @_header_hash, @_mac_key, @_recipient_index)
+    ++@_block_num
+    return payload_text
+
+  _transform : (chunk, encoding, cb) ->
+    if @_header_read
+      super(chunk, encoding, cb)
+    else
+      {_, header_hash, payload_key, _, mac_key, recipient_index} = header.parse_encryption_header_packet(@_decryptor, chunk)
```

Comment:
Are we guaranteed that the first chunk here contains exactly the bytes for the header, rather than part of the header or the-header-plus-part-of-a-payload-packet?